### PR TITLE
Documentation, and inplace support.

### PR DESCRIPTION
**Pull request by ezyang**
This adds the PyTorch API user documentation for Toffee.
To make the example work, I also converted all "inplace"
ops to export out-of-place in Toffee.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: docs/source/toffee.rst [hunk 1]
```diff
@@ -2,6 +2,113 @@ torch.toffee
...
+
+* PyTorch and Caffe2 often have implementations of operators with some
+  numeric differences.  Depending on model structure, these differences
+  may be negligible, but they can also cause major divergences in behavior
+  (especially on untrained models.)  In a future release, we plan to
+  allow Caffe2 to call directly to Torch implementations of operators, to
+  help you smooth over these differences when precision is important,
+  and to also document these differences.
+
+Supported operators
```
Comment (apaszke) [inline docs/source/toffee.rst]: Shouldn't we just send everyone to the official website? We don't want to maintain a separate list ourselves

File: torch/nn/_functions/thnn/pooling.py [hunk 2]
```diff
@@ -12,7 +12,7 @@ class MaxPool1d(Function):
     def primspec(g, input, kernel_size, stride=None, padding=0, dilation=1,
                  ceil_mode=False):
         if ceil_mode:
-            return None
+            raise "ceil_mode not supported in MaxPool1d"
```
Comment (apaszke) [inline torch/nn/_functions/thnn/pooling.py]: RuntimeError!

File: docs/source/toffee.rst (hunk 1 above)
Comment (ezyang) [reply docs/source/toffee.rst]: Yes I'm planning on changing this to a link to the official website when we decide where to put it.

File: docs/source/toffee.rst [hunk 3]
```diff
@@ -2,6 +2,113 @@ torch.toffee
...
+    from torch.autograd import Variable
+    import torch.toffee
+    import torchvision
+
+    dummy_input = Variable(torch.randn(10, 3, 224, 224)).cuda()
+    model = torchvision.models.alexnet(pretrained=True).cuda()
+    torch.toffee.export(model, dummy_input, "alexnet.proto")
+
+The resulting ``alexnet.proto`` is a binary protobuf file which contains both
+the network structure and parameters of AlexNet.  You can verify and inspect
```
Comment (zdevito) [inline docs/source/toffee.rst]: You can pass export any model model whose operators are supported by Toffee, there is nothing special about torchvision's alexnet.

File: docs/source/toffee.rst [hunk 4]
```diff
@@ -2,6 +2,113 @@ torch.toffee
...
+    # ...continuing from above
+    import toffee.backend.c2 as backend
+    import numpy as np
+
+    caffe2_proto = backend.prepare(graph, device="CUDA:0") # or "CPU"
+    outputs = backend.run(caffe2_proto, np.rand(10, 3, 224, 224))
+    print(outputs[0])
+
+In the future, there will be backends for other frameworks as well.
+
```
Comment (zdevito) [inline docs/source/toffee.rst]: How do I get the actual caffe2 protobuf and workspace after prepare? It is likely that if someone is going down this path, they  don't want to run the model through python at all, and are going to need to access the underlying API.

File: docs/source/toffee.rst [hunk 5]
```diff
@@ -2,6 +2,113 @@ torch.toffee
...
+Limitations
+-----------
+
+* The Toffee exporter is a *trace-based* exporter, which means that it
+  operates by executing your model once, and exporting the operators which
+  were actually run during this run.  This means that if your model is
+  dynamic, e.g., changes behavior depending on input data, the export
+  won't be accurate.  Similarly, a trace is likely to be valid only
+  for a specific input size (which is one reason why we require explicit inputs
+  on tracing.)
```
Comment (zdevito) [inline docs/source/toffee.rst]: We suggest examining your model when exporting it and checking that the traced operators look correct.

File: docs/source/toffee.rst [hunk 6]
```diff
@@ -2,6 +2,113 @@ torch.toffee
...
+CPU convolution implementation.  It may take some time to initialize the
+CUDA instance.)::
+
+    from torch.autograd import Variable
+    import torch.toffee
+    import torchvision
+
+    dummy_input = Variable(torch.randn(10, 3, 224, 224)).cuda()
+    model = torchvision.models.alexnet(pretrained=True).cuda()
+    torch.toffee.export(model, dummy_input, "alexnet.proto")
```
Comment (zdevito) [inline docs/source/toffee.rst]: How do I see what I traced? When we were porting models we had to do that frequently. End uses will likely need to do that as well. `verbose=True` would be sufficient.

**Review by zdevito — APPROVED** [review]
I added comments for additional things to add, but this is a good start and should be merged.

File: docs/source/toffee.rst [hunk 7]
```diff
@@ -2,6 +2,113 @@ torch.toffee
...
+
+We plan on expanding support to more operators; RNNs are high on our priority
+list.  The operator set above is sufficient to export the following models:
+
+* AlexNet
+* DCGAN
+* DenseNet
+* Inception (warning: this model is highly sensitive to changes in operator
+  implementation)
+* ResNet50
```
Comment (prigoyal) [inline docs/source/toffee.rst]: you can probably mention ResNet only, the other ResNet like ResNet101 can also be supported very well

**Comment by ezyang** [comment]
I revised based on all the comments.

**Comment by zdevito** [comment]
merging because I am going to build off the pretty printer changes.

**Pull request merged**